src/ZipkinBundle/Components/HttpClient/HttpClient.php

```diff
@@ -0,0 +1,167 @@
+                $spanCustomizer,
+                [$span, 'finish']
+            );
+        } catch (TransportExceptionInterface $e) {
+            // Since response is an asynchronus operation, according to
+            // HttpClientInterface::request, this exception can only happen
+            // if an unsopported option is passed.
+            $span->tag(ERROR, $e->getMessage());
+            $span->finish();
+            throw $e;
+        }
+    }
+
+    /**
+     * buildOnProgress wraps an existing on_progress to listen to the request
+     * completion. on_progress callback will be called on DNS resolution, on
+     * arrival of headers and on completion; it will also be called on upload/download
+     * of data and at least 1/s
+     */
+    private static function buildOnProgress(
+        ?callable $delegateOnProgress,
+        callable $httpParserResponse,
+        SpanCustomizer $spanCustomizer,
+        callable $spanCloser
+    ): callable {
+        return static function (
+            int $dlNow,
+            int $dlSize,
+            array $info
+        ) use (
+            $delegateOnProgress,
+            $httpParserResponse,
+            $spanCustomizer,
+            $spanCloser
+        ): void {
+            if ($delegateOnProgress !== null) {
+                try {
+                    ($delegateOnProgress)($dlNow, $dlSize, $info);
+                } catch (Throwable $e) {
+                    // According to HttpClientInterface, throwing any exceptions
+                    // MUST abort the request, hence we finish the span in here.
+                    $spanCustomizer->tag(ERROR, $e->getMessage());
+                    ($spanCloser)();
+                    throw $e;
+                }
+            }
+
+            if ($info['error'] !== null || $info['http_code'] > 199) {
+                // any of these three conditions represent the finalization of the request
+                ($httpParserResponse)($dlSize, $info, $spanCustomizer);
+                ($spanCloser)();
+            }
+        };
+    }
+
+    /**
+     * unwrapResponses aims to unwrap the wrapped response so the delegate can interpret
+     * it.
+     */
+    private function unwrapResponses($responses): \Generator
+    {
+        foreach ($responses as $response) {
+            yield $response->unwrapResponse();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stream($responses, float $timeout = null): ResponseStreamInterface
+    {
+        if ($responses instanceof Response) {
+            $responses = [$responses];
+        } elseif (!is_iterable($responses)) {
+            throw new TypeError(sprintf(
+                '"%s()" expects parameter 1 to be an iterable of %s objects, "%s" given.',
+                __METHOD__,
+                Response::class,
+                get_class($responses)
+            ));
+        }
+
+        return $this->delegate->stream($this->unwrapResponses($responses), $timeout);
```

Comment:
Check https://github.com/symfony/symfony/pull/36773/files

Reply:
Makes totally sense. Thank you! I just borrowed the implementation design https://github.com/jcchavezs/zipkin-instrumentation-symfony/pull/65/files#diff-68db5e71d0f690cbfaec2c4243986a49R52